### DISCUSSION TITLE
Disabling primeicons has effect on components. Configurable provider?

### DISCUSSION THREAD

**Discussion by djmj** [primefaces/community#100] [PrimeFaces] [answered]
It seems disabling auto inclusion of primeicons using context param _PRIME_ICONS = false_ has effect on primefaces components or premium themes / layouts since some components are using primeicons directly.

This makes primeicons required to use primefaces. Is this intended?

**Examples**

Paginator uses `ui-icon ui-icon-seek-next` but carousel uses `pi pi-chevron-right` instead.

**Background**

We are using font awesome too and would like to stay with this as our primary icon provider since they have a larger variety of icons. For this we can reduce page loading time if we dont  load primeicons CSS file.

In our case there are three icon sets:

- theme specific `ui-icon`
- primeicons `pi`
- font awesome `fa`

**Single icon source / configurable map**

Would be even greater to reduce this to just a single icon resource to be consistent with look and feel across all components and whole website and would not need the icons to be included in the `theme.css`.

Maybe a configurable mapping could be provided as a map in context param that components will user internally.

```
<context-param>
    <param-name>primefaces.ICON_MAP</param-name>
    <param-value>
        ui-icon ui-icon-seek-next=fa fa-chevron-right,
        ui-icon ui-icon-seek-prev=pi pi-chevron-left
    </param-value>
</context-param>
```

For javascript  widget code that map would also need to be provided in js.

**Comment by melloware** [comment, marked as answer]
Yes this is intended. PrimeIcons are the default.  I added this switch because at my clients we generate our own theme using Theme Designer and replaced all icons with FontAwesome.  So I discovered we were loading primeicons.css even though are app was never using it and wasting network resources for no reason. So I added this switch to give the user the ability to NOT include primeicons.css.

You best bet its to do what we did and replaced all `pi` icons with `fa` icons in your theme.scss files.

See this post where i describe what we do with Babylon premium theme: https://forum.primefaces.org/viewtopic.php?f=98&t=57813

**Reply by djmj** [reply]
Thanks for your reply it gave me another idea since we are hosting font awesome on our own server. We are doing this now due to legal reasons in EU where website owners got sued loading google fonts from CDN due to user privacy reasons since lawyers and court did not understand and did not believe the CDN provider (google) statement that no personal IP was logged. Yeah we are always busy with such nonsene here.


Instead of changing the theme.css file and overriding it like you did i will try to edit the font awesome css file instead adding the `pi` rules and deactivate primeicon loading. A quick test in firebug seems to work.

```
.pi, .fa { }

.pi-chevron-right:before, 
.fa-chevron-right:before {
  content: "\f054";
}
```

If primeicons are default in the future will all `ui-icons` class usages will get replaced with their `pi` counterparts?

**Reply by melloware** [reply]
Not sure what the long term plan is but there has been talk of removing all `ui-` based CSS names with `pf-` or `p-` to be more inline with the other prime libraries.
